# ⚡ Project 2: Site Energy Intensity Prediction

🧾**Description:** According to a report issued by the International Energy Agency (IEA), the lifecycle of buildings from construction to demolition was responsible for 37% of global energy-related and process-related CO2 emissions in 2020. Yet it is possible to drastically reduce the energy consumption of buildings by a combination of easy-to-implement fixes and state-of-the-art strategies. 

The dataset consists of building characteristics, weather data for the location of the building, as well as the energy usage for the building, and the given year, measured as Site Energy Usage Intensity (Site EUI). Each row in the data corresponds to a single building observed in a given year.

source of dataset & data dictionary - Click Here  

🧭 **Problem Statement:** You are provided with two datasets: (1) the train_dataset where the observed values of the Site EUI for each row are provided and (2) the x_test dataset the observed values of the Site EUI for each row are removed and provided separately in y_test. Your task is to predict the Site EUI for each row (using the complete training dataset), given the characteristics of the building and the weather data for the location of the building. Use the test sets for validation and testing. 

The target variable is **site_eui** and the evaluation metric is **RMSE** score.


In [6]:
import pandas as pd
from IPython.display import display_html

In [3]:
train = pd.read_csv('data/train_dataset.csv')
test = pd.read_csv('data/x_test.csv')

In [4]:
df_train = train.copy()
df_test = test.copy()

In [5]:
print(df_train.shape)
print(df_test.shape)

(75757, 64)
(9705, 63)


In [8]:
display_html(df_train.head(2), df_train.head(2), df_train.sample(2))

Year_Factor 
 State_Factor 
 building_class 
 facility_type 
 floor_area 
 year_built 
 energy_star_rating 
 ELEVATION 
 january_min_temp 
 january_avg_temp 
 ... 
 days_above_80F 
 days_above_90F 
 days_above_100F 
 days_above_110F 
 direction_max_wind_speed 
 direction_peak_wind_speed 
 max_wind_speed 
 days_with_fog 
 site_eui 
 id 
 
 
 
 
 0 
 1 
 State_1 
 Commercial 
 Grocery_store_or_food_market 
 61242.0 
 1942.0 
 11.0 
 2.4 
 36 
 50.5 
 ... 
 14 
 0 
 0 
 0 
 1.0 
 1.0 
 1.0 
 NaN 
 248.682615 
 0 
 
 
 1 
 1 
 State_1 
 Commercial 
 Warehouse_Distribution_or_Shipping_center 
 274000.0 
 1955.0 
 45.0 
 1.8 
 36 
 50.5 
 ... 
 14 
 0 
 0 
 0 
 1.0 
 NaN 
 1.0 
 12.0 
 26.500150 
 1 
 
 
 
 2 rows × 64 columns

Year_Factor 
 State_Factor 
 building_class 
 facility_type 
 floor_area 
 year_built 
 energy_star_rating 
 ELEVATION 
 january_min_temp 
 january_avg_temp 
 ... 
 days_above_80F 
 days_above_90F 
 days_above_100F 
 days_above_110F 
 direction_max_wind_speed 
 direction_peak_wind_speed 
 max_wind_speed 
 days_with_fog 
 site_eui 
 id 
 
 
 
 
 0 
 1 
 State_1 
 Commercial 
 Grocery_store_or_food_market 
 61242.0 
 1942.0 
 11.0 
 2.4 
 36 
 50.5 
 ... 
 14 
 0 
 0 
 0 
 1.0 
 1.0 
 1.0 
 NaN 
 248.682615 
 0 
 
 
 1 
 1 
 State_1 
 Commercial 
 Warehouse_Distribution_or_Shipping_center 
 274000.0 
 1955.0 
 45.0 
 1.8 
 36 
 50.5 
 ... 
 14 
 0 
 0 
 0 
 1.0 
 NaN 
 1.0 
 12.0 
 26.500150 
 1 
 
 
 
 2 rows × 64 columns

Year_Factor 
 State_Factor 
 building_class 
 facility_type 
 floor_area 
 year_built 
 energy_star_rating 
 ELEVATION 
 january_min_temp 
 january_avg_temp 
 ... 
 days_above_80F 
 days_above_90F 
 days_above_100F 
 days_above_110F 
 direction_max_wind_speed 
 direction_peak_wind_speed 
 max_wind_speed 
 days_with_fog 
 site_eui 
 id 
 
 
 
 
 73624 
 6 
 State_11 
 Commercial 
 Office_Uncategorized 
 39760.0 
 1925.0 
 92.0 
 26.5 
 28 
 43.451613 
 ... 
 25 
 3 
 0 
 0 
 NaN 
 NaN 
 NaN 
 NaN 
 62.178272 
 73624 
 
 
 41686 
 4 
 State_6 
 Residential 
 Multifamily_Uncategorized 
 66874.0 
 2008.0 
 47.0 
 42.7 
 4 
 28.596774 
 ... 
 75 
 4 
 0 
 0 
 1.0 
 1.0 
 1.0 
 NaN 
 87.126668 
 41686 
 
 
 
 2 rows × 64 columns

In [9]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75757 entries, 0 to 75756
Data columns (total 64 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year_Factor                75757 non-null  int64  
 1   State_Factor               75757 non-null  object 
 2   building_class             75757 non-null  object 
 3   facility_type              75757 non-null  object 
 4   floor_area                 75757 non-null  float64
 5   year_built                 73920 non-null  float64
 6   energy_star_rating         49048 non-null  float64
 7   ELEVATION                  75757 non-null  float64
 8   january_min_temp           75757 non-null  int64  
 9   january_avg_temp           75757 non-null  float64
 10  january_max_temp           75757 non-null  int64  
 11  february_min_temp          75757 non-null  int64  
 12  february_avg_temp          75757 non-null  float64
 13  february_max_temp          75757 non-null  int

In [10]:
df_train.rename(columns={
    'Year_Factor': 'year_factor', 
    'State_Factor': 'state_factor',
    'ELEVATION': 'elevation',
    'id': 'building_id'
}, inplace=True)

df_test.rename(columns={
    'Year_Factor': 'year_factor', 
    'State_Factor': 'state_factor',
    'ELEVATION': 'elevation',
    'id': 'building_id'
}, inplace=True)

# df.columns = df.columns.values.lower()

# EDA

year_Factor: anonymized year in which the weather and energy usage factors were observed  

state_Factor: anonymized state in which the building is located  

building_class: building classification  

facility_type: building usage type  

floor_area: floor area (in square feet) of the building  

year_built: year in which the building was constructed  

energy_star_rating: the energy star rating of the building  

elevation: elevation of the building location  

january_min_temp: minimum temperature in January (in Fahrenheit) at the location of the building  

january_avg_temp: average temperature in January (in Fahrenheit) at the location of the building  

january_max_temp: maximum temperature in January (in Fahrenheit) at the location of the building  

[Similarly for all other months]

cooling_degree_days: cooling degree day for a given day is the number of degrees where the daily average temperature   
exceeds 65 degrees Fahrenheit. Each month is summed to produce an annual total at the location of the building.  

heating_degree_days: heating degree day for a given day is the number of degrees where the daily average temperature   falls under 65 degrees Fahrenheit. Each month is summed to produce an annual total at the location of the building.  

precipitation_inches: annual precipitation in inches at the location of the building  

snowfall_inches: annual snowfall in inches at the location of the building  

snowdepth_inches: annual snow depth in inches at the location of the building  

avg_temp: average temperature over a year at the location of the building  

days_below_30F: total number of days below 30 degrees Fahrenheit at the location of the building  

days_below_20F: total number of days below 20 degrees Fahrenheit at the location of the building  

days_below_10F: total number of days below 10 degrees Fahrenheit at the location of the building  

days_below_0F: total number of days below 0 degrees Fahrenheit at the location of the building  

days_above_80F: total number of days above 80 degrees Fahrenheit at the location of the building  

days_above_90F: total number of days above 90 degrees Fahrenheit at the location of the building  

days_above_100F: total number of days above 100 degrees Fahrenheit at the location of the building  

days_above_110F: total number of days above 110 degrees Fahrenheit at the location of the building  

direction_max_wind_speed: wind direction for maximum wind speed at the location of the building. Given in 360-degree compass point directions (e.g. 360 = north, 180 = south, etc.).  

direction_peak_wind_speed: wind direction for peak wind gust speed at the location of the building. Given in 360-degree compass point directions (e.g. 360 = north, 180 = south, etc.).  

max_wind_speed: maximum wind speed at the location of the building  

days_with_fog: number of days with fog at the location of the building  

building_id: building id  

#### Target

**site_eui:** Site Energy Usage Intensity is the amount of heat and electricity consumed by a building as reflected in utility bills 

In [12]:
numerical_feature = [feature for feature in df_train.columns if df_train[feature].dtypes != 'O']
discrete_feature=[feature for feature in numerical_feature if len(df_train[feature].unique())<25]
continuous_feature = [feature for feature in numerical_feature if feature not in discrete_feature]
categorical_feature = [feature for feature in df_train.columns if feature not in numerical_feature]
print("Numerical Features Count {}".format(len(numerical_feature)))
print("Discrete feature Count {}".format(len(discrete_feature)))
print("Continuous feature Count {}".format(len(continuous_feature)))
print("Categorical feature Count {}".format(len(categorical_feature)))

Numerical Features Count 61
Discrete feature Count 6
Continuous feature Count 55
Categorical feature Count 3


In [13]:
print(df_train.duplicated().sum())
print(df_test.duplicated().sum())

0
0


In [16]:
print([col for col in df_train if df_train[col].nunique()==1])
print([col for col in df_test if df_test[col].nunique()==1])

[]
['year_factor', 'days_above_110F']


In [17]:
def missing_values_table(df):
        # Total missing values by column
        mis_val = df.isnull().sum()
        
        # Percentage of missing values by column
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        
        # build a table with the thw columns
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        
        # Rename the columns
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        
        # Sort the table by percentage of missing descending
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        
        # Print some summary information
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        
        # Return the dataframe with missing information
        return mis_val_table_ren_columns

In [18]:
print("Train set columns with null values: ")
print(list(df_train.columns[df_train.isnull().any()]))
print('===========================================')
# Missing values for training data
missing_values_train = missing_values_table(df_train)
missing_values_train[:20].style.background_gradient(cmap='Reds')

Train set columns with null values: 
['year_built', 'energy_star_rating', 'direction_max_wind_speed', 'direction_peak_wind_speed', 'max_wind_speed', 'days_with_fog']
Your selected dataframe has 64 columns.
There are 6 columns that have missing values.


In [19]:
print("Test set columns with null values: ")
print(list(df_test.columns[df_test.isnull().any()]))
print('===========================================')
# Missing values for test data
missing_values_test = missing_values_table(df_test)
missing_values_test[:20].style.background_gradient(cmap='Reds')

Test set columns with null values: 
['year_built', 'energy_star_rating', 'direction_max_wind_speed', 'direction_peak_wind_speed', 'max_wind_speed', 'days_with_fog']
Your selected dataframe has 63 columns.
There are 6 columns that have missing values.


<div style="border-radius:10px; border:#91C7B1 solid; padding: 15px; background-color: #f1f7ed; font-size:100%; text-align:left">
<h3 align="left"><font color='#91C7B1'>💡 Observations:</font></h3>

1) `year_built`, `energy_star_rating`, `direction_max_wind_speed`, `direction_peak_wind_speed`, `max_wind_speed`, `days_with_fog` variables have `NA` values.   
2) Data contains 75757 samples and 64 features. We have 3 categorical and rest numerical features. Target is numerical.   
3) There are no duplicates in individual dataset, however there might be duplicates in the combined version   
4) there are no constant columns in train dataset, but test dataset has 2 constant columns   